runtime/runtime/src/bandwidth_scheduler/mod.rs

```diff
@@ -38,55 +48,89 @@ pub fn run_bandwidth_scheduler(
         Some(prev_state) => prev_state,
         None => {
             tracing::debug!(target: "runtime", "Bandwidth scheduler state not found - initializing");
-            BandwidthSchedulerState { mock_data: [0; 32] }
+            BandwidthSchedulerState {
+                link_allowances: Vec::new(),
+                sanity_check_hash: CryptoHash::default(),
+            }
         }
     };
 
-    let scheduler_state_hash: CryptoHash = hash(&borsh::to_vec(&scheduler_state).unwrap());
+    // Prepare the status info for each shard based on congestion info.
+    let mut shards_status: BTreeMap<ShardId, ShardStatus> = BTreeMap::new();
```

Comment:
I have recently removed Ord and PartialOrd from ShardIds. The usage of BTreeMap on ShardIds led to conflating those with ShardIndexes. In principle the ShardIds should be considered arbitrary ids and as such they shouldn't be sortable. You can always use ShardIndexes if you need this functionality.

Reply:
Ah it seems I failed to fully remove ord from ShardId (the failed attempt here didn't get rid of the derives.. #12436). Still it would be nice to not depend on it but feel free to leave it as is and I can follow up later.

Reply:
> Also FYI I'm adding the shard layout to the runtime via the EpochInfoProvider in another PR.

Nice that will be very useful!

I took a look at https://github.com/near/nearcore/pull/12538 and I see that `EpochInfoProvider` has knowledge about the current shard layout. It would be useful to also have information about the shard layout that the receivers will use, so that we could get the lists of sender and receiver shards. Is that hard to do? I wonder if it's possible to know what the shard layout will be at the next block, is there any trouble with missing blocks/epoch boundaries or something...

Reply:
I think in principle it should be doable, something like:
```rust
fn shard_layout(&self, epoch_id: &EpochId) -> Result<ShardLayout, EpochError> {
    let epoch_manager = self.read();
    epoch_manager.get_shard_layout(epoch_id)
}

fn next_block_shard_layout(
    &self,
    current_block_hash: &CryptoHash,
) -> Result<ShardLayout, EpochError> {
    let epoch_manager = self.read();
    let next_block_epoch_id = epoch_manager.get_epoch_id_from_prev_block(current_block_hash)?;
    epoch_manager.get_shard_layout(&next_block_epoch_id)
}
...
let sender_shards: Vec<ShardId> =
    epoch_info_provider.shard_layout(&apply_state.epoch_id).unwrap().shard_ids().collect();
let receiver_shards: Vec<ShardId> = epoch_info_provider
    .next_block_shard_layout(&apply_state.block_hash)
    .unwrap()
    .shard_ids()
    .collect();
```

But I'm not gonna do that in this PR, I'd classify that as a possible future improvement.